### PR TITLE
Allow specifying target environment variables as disabled

### PR DESCRIPTION
**Pull request by lo1tuma**
As discussed [here](https://github.com/tuist/tuist/discussions/3257) this change allows to control scheme environment variables if they should be enabled or not if the scheme action runs.

### CODE REVIEW

File: Sources/ProjectDescription/Arguments.swift [hunk 1]
```diff
@@ -2,14 +2,14 @@ import Foundation
...
 public struct Arguments: Equatable, Codable {
-    public let environment: [String: String]
+    public let environmentVariables: [EnvironmentVariable]
     public let launchArguments: [LaunchArgument]
 
     public init(
-        environment: [String: String] = [:],
+        environmentVariables: [EnvironmentVariable] = [],
         launchArguments: [LaunchArgument] = []
     ) {
```
[the remark this answers is not in the document]
Comment (danyf90) [reply Sources/ProjectDescription/Arguments.swift:11]: Also:
- I would keep it as a dictionary
- the `ProjectDescription.EnvironmentVariable` should not have the name variable
- the `ProjectDescription.EnvironmentVariable` should be `ExpressibleByStringLiteral`

**Review by danyf90 — CHANGES_REQUESTED** [review]

**Comment by danyf90** [comment]
@allcontributors add @lo1tuma for code

**Comment by lo1tuma** [comment]
@danyf90 thanks for your feedback. I’ve finally had time to continue working on this and addressed all of your comments.

File: Sources/ProjectDescription/Arguments.swift [hunk 2]
```diff
@@ -2,14 +2,30 @@ import Foundation
 
 /// A collection of arguments and environment variables.
 public struct Arguments: Equatable, Codable {
-    public let environment: [String: String]
+    public let environmentVariables: [String: EnvironmentVariable]
     public let launchArguments: [LaunchArgument]
 
+    @available(*, deprecated, message: "please use environmentVariables instead")
     public init(
-        environment: [String: String] = [:],
```
Comment (pepicrft) [inline Sources/ProjectDescription/Arguments.swift:9]: This seems to be a breaking change. I'd keep this as it is with the deprecated warning, and then we'll remove it when we have the opportunity to do a major release.

File: Sources/ProjectDescription/EnvironmentVariable.swift [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    public let isEnabled: Bool
+
+    // MARK: - Init
+
+    public init(value: String, isEnabled: Bool) {
+        self.value = value
+        self.isEnabled = isEnabled
+    }
+
+    public init(stringLiteral value: String) {
```
Comment (pepicrft) [inline Sources/ProjectDescription/EnvironmentVariable.swift:19]: I'd add documentation here to indicate that it initialises it being enabled by default.

**Review by pepicrft — APPROVED** [review]

File: Sources/ProjectDescription/Arguments.swift (hunk 2 above)
Comment (danyf90) [reply Sources/ProjectDescription/Arguments.swift:9]: @lo1tuma please add back the default value 🙏

File: Sources/TuistGraph/Models/EnvironmentVariable.swift [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    public init(value: String, isEnabled: Bool) {
+        self.value = value
+        self.isEnabled = isEnabled
+    }
+
+    public init(stringLiteral value: String) {
+        self.value = value
+        isEnabled = true
+    }
```
Comment (danyf90) [inline Sources/TuistGraph/Models/EnvironmentVariable.swift:22]: nit: No need for this as it's an internal type, better to make sure we use the proper init

**Review by danyf90 — APPROVED** [review]

File: Sources/ProjectDescription/Arguments.swift (hunk 2 above)
Comment (lo1tuma) [reply Sources/ProjectDescription/Arguments.swift:9]: Good catch. There are multiple initializers, and this only affects calling `Arguments()` without any parameters. I’ve added the default value to second initializer to bring this functionality back and not causing the deprecation warning to be triggered for calling `Arguments()`.

File: Sources/ProjectDescription/EnvironmentVariable.swift (hunk 3 above)
Comment (lo1tuma) [reply Sources/ProjectDescription/EnvironmentVariable.swift:19]: That’s the same behavior as it has been before. But I’m happy to document this. Please let me know where and how I can add the documentation.

File: Sources/TuistGraph/Models/EnvironmentVariable.swift (hunk 4 above)
Comment (lo1tuma) [reply Sources/TuistGraph/Models/EnvironmentVariable.swift:22]: Done

**Review by danyf90 — APPROVED** [review]

**Comment by pepicrft** [comment]
I fixed compilation and style issues and rebased from `main`. I'll merge when CI is green.

**Pull request merged**